qiskit/algorithms/optimizers/gradient_descent.py

```diff
@@ -149,60 +227,99 @@ def settings(self) -> Dict[str, Any]:
+        else:
+            grad = self._state.jac(ask_object.x_jac)
+            self._state.njev += 1
+
+        return TellObject(eval_jac=grad)
+
+    def create_result(self) -> OptimizerResult:
+        """Creates a result of the optimization process using the values from
+        :class:`~.GradientDescentState`.
+        """
+        result = OptimizerResult()
+        result.x = self._state.x
+        result.fun = self._state.fun(self._state.x)
+        result.nfev = self._state.nfev
+        result.njev = self._state.njev
+        result.nit = self._state.nit
+        return result
+
+    def initialize(
         self,
         fun: Callable[[POINT], float],
         x0: POINT,
         jac: Optional[Callable[[POINT], POINT]] = None,
         bounds: Optional[List[Tuple[float, float]]] = None,
-    ) -> OptimizerResult:
-        # set learning rate
+    ) -> None:
+        """
+        This method will initialize the state of the optimizer so that an optimization can be performed.
+        It will always setup the initial point and will restart the counter for function evaluations.
+        This method is left blank because every optimizer has a different kind of state.
+        """
         if isinstance(self.learning_rate, float):
             eta = constant(self.learning_rate)
         else:
             eta = self.learning_rate()
 
-        if jac is None:
-            eps = 0.01 if self.perturbation is None else self.perturbation
-            jac = partial(
-                Optimizer.gradient_num_diff,
-                f=fun,
-                epsilon=eps,
-                max_evals_grouped=self._max_evals_grouped,
-            )
-
-        # prepare some initials
-        x = np.asarray(x0)
-        nfevs = 0
-
-        for _ in range(1, self.maxiter + 1):
-            # compute update -- gradient evaluation counts as one function evaluation
-            update = jac(x)
-            nfevs += 1
-
-            # compute next parameter value
-            x_next = x - next(eta) * update
-
-            # send information to callback
-            stepsize = np.linalg.norm(update)
-            if self.callback is not None:
-                self.callback(nfevs, x_next, fun(x_next), stepsize)
-
-            # update parameters
-            x = x_next
-
-            # check termination
-            if stepsize < self.tol:
-                break
-
-        # TODO the optimizer result should contain the number of gradient evaluations,
-        # if the gradient is passed
-        result = OptimizerResult()
-        result.x = x
-        result.fun = fun(x)
-        result.nfev = nfevs
-
-        return result
+        self._state = GradientDescentState(
+            fun=fun,
+            jac=jac,
+            x=np.asarray(x0),
+            nit=0,
+            nfev=0,
+            njev=0,
+            eta=eta,
+            stepsize=None,
+        )
+
+    def continue_condition(self) -> bool:
```

Comment:
I commented about a method above and being public - some like this seem to be internal operation of the optimizer so being part of public API does not seem 'right' to me.

Reply:
I believe the user should be able to easily access continue_condition in case he wants to create his own optimization loop. Of course he could check the condition by hand, and for gradient descent it wouldn't be too bad, but for other optimizers it will be a headache to redo the stopping criteria.